server/resp/src/main/java/org/infinispan/server/resp/SubscriberHandler.java

```diff
@@ -170,29 +167,20 @@ public CompletionStage<RespRequestHandler> unsubscribeAll(ChannelHandlerContext
    }
 
    public CompletionStage<RespRequestHandler> sendSubscriptions(ChannelHandlerContext ctx, CompletionStage<Void> stageToWaitFor,
-         Collection<byte[]> keyChannels, boolean subscribeOrUnsubscribe) {
+                                                                Collection<byte[]> keyChannels, boolean isSubscribe) {
       return stageToReturn(stageToWaitFor, ctx, (__, alloc) -> {
          assert ctx.executor().inEventLoop();
 
          Long counter = ctx.channel().attr(SUBSCRIPTIONS_COUNTER).get();
          if (counter == null) counter = 0L;
 
+         // PubSub events require the object type to be a bulk string.
+         byte[] type = isSubscribe ? PubSubEvents.SUBSCRIBE : PubSubEvents.UNSUBSCRIBE;
          for (byte[] keyChannel : keyChannels) {
-            String initialCharSeq = subscribeOrUnsubscribe ? "*3\r\n$9\r\nsubscribe\r\n$" : "*3\r\n$11\r\nunsubscribe\r\n$";
-            counter = Math.max(0, counter + (subscribeOrUnsubscribe ? 1 : -1));
-
-            // Length of string (all ascii so 1 byte per) + (log10 + 1 = sizes of number as char in bytes) + \r\n + bytes themselves + \r\n + : + countSize + \r\n
-            int countSize = counter == 0 ? 1 : (int) Math.log10(counter) + 1;
-            int sizeRequired = initialCharSeq.length() + (int) Math.log10(keyChannel.length) + 1 + 2 + keyChannel.length + 2 + 1 + countSize + 2;
-            ByteBuf subscribeBuffer = alloc.apply(sizeRequired);
-            int initialPos = subscribeBuffer.writerIndex();
-            subscribeBuffer.writeCharSequence(initialCharSeq + keyChannel.length + CRLF_STRING, CharsetUtil.US_ASCII);
-            subscribeBuffer.writeBytes(keyChannel);
-            subscribeBuffer.writeBytes(CRLF);
-            subscribeBuffer.writeByte(':');
-            ByteBufferUtils.setIntChars(counter, countSize, subscribeBuffer);
-            subscribeBuffer.writeBytes(CRLF);
-            assert subscribeBuffer.writerIndex() - initialPos == sizeRequired;
+            counter = Math.max(0, counter + (isSubscribe ? 1 : -1));
+            List<Object> reply = List.of(type, keyChannel, counter);
+            // Mixed types in array.
+            Resp3Response.array(reply, alloc);
```

Comment:
I am not sure about this, tbh. I think places that use array where it is a known number of elements that have discrete types we should instead write the array length in this method and then the elements. This way no special serializer lookup is required.

Reply:
Also reworked this. Anything that uses a heterogeneous structures (map, set, list), should provide a serializer. It makes it more restrictive but should avoid cases of iterating in search of all the serializers.